src/usdb_syncer/pdf.py

```diff
@@ -1,40 +1,217 @@
+)
 
-def generate_song_pdf(songs: Iterable[SongId], path: str) -> None:
-    document: dict[str, Any] = {}
-    document["style"] = {"margin_bottom": 15, "text_align": "j"}
-    document["formats"] = {"url": {"c": "blue", "u": 1}, "title": {"b": 1, "s": 13}}
-    document["sections"] = []
-    section1: dict[str, list[Any]] = {}
-    document["sections"].append(section1)
-    content1: list[Any] = []
-    section1["content"] = content1
-    date = datetime.datetime.now()
-    content1.append(
-        {
-            ".": f"Songlist ({date:%Y-%m-%d})",
-            "style": "title",
-            "label": "title1",
-            "outline": {"level": 1, "text": "A different title 1"},
-        }
+
+class Bookmark(Flowable):
+    """Bookmark flowable."""
+
+    def __init__(self, key: str, title: str, level: int = 0):
+        super().__init__()
+        self.key = key
+        self.title = title
+        self.level = level
+
+    def draw(self) -> None:
+        self.canv.bookmarkPage(self.key)
+        self.canv.addOutlineEntry(self.title, self.key, level=self.level, closed=False)
+
+
+def generate_report_pdf(
+    *,
+    songs: Iterable[SongId],
+    path: str,
+    size: ReportPagesize = ReportPagesize.A4,
+    orientation: ReportOrientation = ReportOrientation.PORTRAIT,
+    margin: int = 20,
+    column_count: int = 1,
+    base_font_size: int = 10,
+    optional_info: list[Column] | None = None,
+) -> None:
+    optional_info = optional_info or []
+    pagesize = _get_pagesize(size, orientation)
+
+    doc: BaseDocTemplate = BaseDocTemplate(
+        path,
+        pagesize=pagesize,
+        leftMargin=margin * mm,
+        rightMargin=margin * mm,
+        topMargin=margin * mm,
+        bottomMargin=margin * mm,
+    )
+
+    custom_styles = _create_paragraph_styles(base_font_size)
+    content = _build_pdf_content(songs, base_font_size, custom_styles, optional_info)
+
+    gutter = margin * mm * 0.5
+    frames = _create_column_frames(
+        pagesize, margin=margin * mm, gutter=gutter, column_count=column_count
+    )
+    template = PageTemplate(
+        id="with_page_number", frames=frames, onPage=_add_page_number
+    )
+    doc.addPageTemplates([template])
+    doc.build(content)
+
+
+def _get_pagesize(
+    size: ReportPagesize, orientation: ReportOrientation
+) -> tuple[float, float]:
+    match size:
+        case ReportPagesize.A3:
+            pagesize = A3
+        case ReportPagesize.A4:
+            pagesize = A4
+        case ReportPagesize.A5:
+            pagesize = A5
+        case ReportPagesize.LEGAL:
+            pagesize = LEGAL
+        case ReportPagesize.LETTER:
+            pagesize = LETTER
+        case _ as unreachable:
+            assert_never(unreachable)
+    return (
+        landscape(pagesize) if orientation == ReportOrientation.LANDSCAPE else pagesize
     )
 
+
+def _create_paragraph_styles(base_font_size: int) -> dict[str, ParagraphStyle]:
+    return {
+        "Initial": ParagraphStyle(
```

Comment:
Should be type-safe, e.g. by using a dataclass instead of a dict with magic keys.